pkg/serviceaccounttoken/secret.go

```diff
@@ -5,41 +5,61 @@ import (
 	"fmt"
 	"time"
 
+	corecontrollers "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
 	"github.com/sirupsen/logrus"
 	v1 "k8s.io/api/core/v1"
-	apierror "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/kubernetes"
 )
 
-const serviceAccountSecretAnnotation = "kubernetes.io/service-account.name"
+const (
+	// ServiceAccountSecretLabel is the label used to search for the secret belonging to a service account.
+	ServiceAccountSecretLabel = "cattle.io/service-account.name"
 
-// secretGetter is an abstraction over any kind of secret getter.
+	serviceAccountSecretAnnotation = "kubernetes.io/service-account.name"
+)
+
+// secretLister is an abstraction over any kind of secret lister.
 // The caller can use any cache or client it has available, whether that is from norman, wrangler, or client-go,
 // as long as it can wrap it in a simplified lambda with this signature.
-type secretGetter func(namespace, name string) (*v1.Secret, error)
+type secretLister func(namespace string, selector labels.Selector) ([]*v1.Secret, error)
 
 // EnsureSecretForServiceAccount gets or creates a service account token Secret for the provided Service Account.
 // For k8s <1.24, the secret is automatically generated for the service account. For >=1.24, we need to generate it explicitly.
-func EnsureSecretForServiceAccount(ctx context.Context, secretGetter secretGetter, clientSet kubernetes.Interface, sa *v1.ServiceAccount) (*v1.Secret, error) {
-	if secretGetter == nil {
-		secretGetter = func(namespace, name string) (*v1.Secret, error) {
-			return clientSet.CoreV1().Secrets(namespace).Get(ctx, name, metav1.GetOptions{})
-		}
-	}
+func EnsureSecretForServiceAccount(ctx context.Context, secretsCache corecontrollers.SecretCache, clientSet kubernetes.Interface, sa *v1.ServiceAccount) (*v1.Secret, error) {
 	if sa == nil {
 		return nil, fmt.Errorf("could not ensure secret for invalid service account")
 	}
 	secretClient := clientSet.CoreV1().Secrets(sa.Namespace)
-	saClient := clientSet.CoreV1().ServiceAccounts(sa.Namespace)
-	secretName := ServiceAccountSecretName(sa)
-	var secret *v1.Secret
-	var err error
-	if secretName != "" {
-		secret, err = secretGetter(sa.Namespace, secretName)
-		if err != nil && !apierror.IsNotFound(err) {
-			return nil, fmt.Errorf("error ensuring secret for service account [%s:%s]: %w", sa.Namespace, sa.Name, err)
+	var secretLister secretLister
+	if secretsCache != nil {
+		secretLister = secretsCache.List
+	} else {
+		secretLister = func(_ string, selector labels.Selector) ([]*v1.Secret, error) {
+			secretList, err := secretClient.List(ctx, metav1.ListOptions{
```

Comment:
Rather than using a Label selector, we could use an indexer which maps a given secret to its `kubernetes.io/service-account.name` value. This has the following advantages:
- It will allow us to identify both new and old secrets
- It would avoid needing to process returned values to figure out if they are really service account tokens
- All this happens in memory, so there's no cases where we need to reach out to the API server.

This would also allow us to delete the old tokens when ensuring, since the indexer (unlike the selector) will identify tokens made before this change and after this change. I would be a bit worried about the edge cases here (namely if another service has persisted this token outside of k8s and is relying on that value still being valid, this plan would break that).

Reply:
An indexer would only work if we have a controller initialized, which is not always the case as described above.